packages/evm/src/eof/verify.ts

```diff
@@ -0,0 +1,519 @@
+            const expectedStack = currentOutputs + targetInputs - targetOutputs
+            if (!(minStackCurrent === maxStackCurrent && maxStackCurrent === expectedStack)) {
+              validationError(EOFError.InvalidStackHeight)
+            }
+          }
+          if (
+            maxStackCurrent + container.body.typeSections[target].maxStackHeight - targetInputs >
+            1024
+          ) {
+            //console.log(maxStackCurrent, targetOutputs, targetInputs, targetNonReturning)
+            validationError(EOFError.StackOverflow)
+          }
+        }
+      } else if (opcode === 0xe4) {
+        // RETF
+        // Stack height must match the outputs of current code section
+        const outputs = container.body.typeSections[codeSection].outputs
+        if (!(minStackCurrent === maxStackCurrent && maxStackCurrent === outputs)) {
+          validationError(EOFError.InvalidStackHeight)
+        }
+      } else if (opcode === 0xe6) {
+        // DUPN
+        const toDup = code[ptr + 1]
+        if (toDup + 1 > minStackCurrent) {
+          validationError(EOFError.StackUnderflow)
+        }
+      } else if (opcode === 0xe7) {
+        // SWAPN
+        const toSwap = code[ptr + 1]
+        // TODO: EVMONEs test wants this to be `toSwap + 2`, but that seems to be incorrect
+        // Will keep `toSwap + 1` for now
+        if (toSwap + 1 > minStackCurrent) {
+          validationError(EOFError.StackUnderflow)
+        }
+      } else if (opcode === 0xe8) {
+        // EXCHANGE
+        const exchangeRaw = code[ptr + 1]
+        const n = (exchangeRaw >> 4) + 1
+        const m = (exchangeRaw & 0x0f) + 1
+        if (n + m + 1 > minStackCurrent) {
+          validationError(EOFError.StackUnderflow)
+        }
+      } else if (opcode === 0xec) {
+        // EOFCREATE
+        const target = code[ptr + 1]
+        if (target >= container.header.containerSizes.length) {
+          validationError(EOFError.InvalidEOFCreateTarget)
+        }
+        if (containerTypeMap.has(target)) {
+          if (containerTypeMap.get(target) !== ContainerSectionType.InitCode) {
+            validationError(EOFError.ContainerDoubleType)
+          }
+        }
+        containerTypeMap.set(target, ContainerSectionType.InitCode)
+      } else if (opcode === 0xee) {
+        // RETURNCONTRACT
+
+        if (mode !== ContainerSectionType.InitCode) {
+          validationError(EOFError.ContainerTypeError)
+        }
+
+        const target = code[ptr + 1]
+        if (target >= container.header.containerSizes.length) {
+          validationError(EOFError.InvalidRETURNContractTarget)
+        }
+        if (containerTypeMap.has(target)) {
+          if (containerTypeMap.get(target) !== ContainerSectionType.DeploymentCode) {
+            validationError(EOFError.ContainerDoubleType)
+          }
+        }
+        containerTypeMap.set(target, ContainerSectionType.DeploymentCode)
+      } else if (opcode === 0xd1) {
+        // DATALOADN
```

Comment:
We cannot do this, in this context we do not have access to the stack. The value on the stack could be anything. We thus have to verify this at runtime.

Side-note: this could be made more clear in the EIP itself, the difference between container validation and runtime validation!